File: sunflower-core/src/main/java/org/tdf/sunflower/Stop.java
package org.tdf.sunflower;

import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.tdf.sunflower.facade.DatabaseStoreFactory;
import org.tdf.sunflower.facade.Miner;
import org.tdf.sunflower.net.PeerServer;

import javax.annotation.PreDestroy;

// resource cleaner
@Component
@Slf4j(topic = "stop")
public class Stop {

    @Autowired
    private Miner miner;

    @Autowired
    private PeerServer peerServer;

    @Autowired
    private DatabaseStoreFactory factory;


    public Stop() {
    }

    @PreDestroy
    public void onDestroy() throws Exception {
        log.info("clean resources before exit program...");
        miner.stop();
        peerServer.stop();
        factory.cleanup();
    }
}

File: sunflower-core/src/main/java/org/tdf/sunflower/Stop.kt
package org.tdf.sunflower

import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.tdf.sunflower.facade.DatabaseStoreFactory
import org.tdf.sunflower.facade.Miner
import org.tdf.sunflower.net.PeerServer
import javax.annotation.PreDestroy

// resource cleaner
@Component
class Stop(
    private val miner: Miner,
    private val peerServer: PeerServer,
    private val factory: DatabaseStoreFactory
) {
    @PreDestroy
    @Throws(Exception::class)
    fun onDestroy() {
        log.info("clean resources before exit program...")
        miner.stop()
        peerServer.stop()
        factory.cleanup()
    }

    companion object {
        private val log = LoggerFactory.getLogger("stop")
    }
}